fix(npm): various fixes to the cli npm logic

kbrandwijk opened the pull request:
Fixes 4 open issues regarding npm logic. Please see individual commits for details.

Fixes #60, fixes #107, fixes #18, fixes #79.

pvdlg commented:
I didn't look in detail, but for info we are using one commit per feature/fix and not a work log type. So we get a clean commit history.

So for example commit like `style: linting` should be squashed with the commit in which the linting error were introduced.

See http://www.bitsnbites.eu/git-history-work-log-vs-recipe/ for more details

Also for simplicity, whenever possible, it's better to open multiple more specific so it's easier to review (maybe it's not relevant regarding this PR)

kbrandwijk commented:
To be honest, if I create a new branch to fix some issues, I'm not going to submit 4 PR's and create 4 new branches in 2 hours time. I changed the PR description after I discovered I forgot to lint, because the argument of a clean commit history didn't apply anymore because of that. The title and description are now updated to allow squashing.

gr2m commented on src/lib/npm.js:
```diff
@@ -23,9 +23,17 @@ async function getNpmToken({npm, options}) {
+          .catch(err => ({ err }))
+      }
+    }
+  )
```
👍

gr2m commented on src/lib/npm.js:
```diff
@@ -42,13 +42,27 @@ async function getNpmToken({npm, options}) {
+
+  return conf.get('registry') || 'https://registry.npmjs.org/'
+
+}
```
code looks good, but moving forward we really need to improve the test / code coverage situation :)

kbrandwijk replied:
Agreed

gr2m commented on src/lib/npm.js:
```diff
@@ -67,6 +69,14 @@ module.exports = async function(pkg, info) {
+      type: 'list',
+      name: 'authmethod',
+      message: 'Which authentication method is this npm registry using?',
+      choices: [{name: 'Token based', value: 'token'}, {name: 'Legacy', value: 'legacy'}],
```
maybe clarify what `Legacy` means, like `'Legacy (username & password)`'? Maybe it’s clear enough, I can’t tell, as I’ve never run into it myself

pvdlg replied:
If we clarify what legacy is we should mention username, password and email, as email is required.

kbrandwijk replied:
Updated

pvdlg commented on src/lib/npm.js:
```diff
@@ -42,14 +41,31 @@ async function getNpmToken({npm, options}) {
   log.info('Successfully created npm token.');
 }
 
+function getRegistry(pkg, conf) {
```
Where do we use that in semantic-release to retrieve the registry?

AFAK we use https://github.com/semantic-release/npm/blob/master/lib/get-registry.js and its the only place where we read the registry

pvdlg replied:
That's super old, it's not in the code base anymore.
We should use https://github.com/semantic-release/npm/blob/master/lib/get-registry.js for consistency.

But that can be done in a different PR. Let me know if you prefer to do it here, otherwise I can merge this PR

kbrandwijk replied:
There's no runtime dep on semantic-release yet. Is there any way to reuse this, without copying?

pvdlg replied:
It's in a plugin so not easily.
It's only 2 lines of code and it uses [registry-auth-token](https://github.com/rexxars/registry-auth-token), so I guess it ok.

pvdlg commented on src/lib/npm.js:
```diff
@@ -25,22 +22,65 @@ async function getNpmToken({npm, options}) {
-        .catch(err => ({err}));
-    }
+
+  const {err, token} = await new Promise(resolve => {
```
Why changing to a callback call here?
You could use `async/await` instead:
```js
let err;
let token;
try {
({err, token} = await client.requestAsync(uri, {method: 'PUT', body}));
} catch (err) {
if (err.code === 'E409') {
....
} else if (....) {
....
}

```

kbrandwijk replied:
The reason for using a callback is that Promises (either directly or through promisify) don't support getting multiple arguments back on failure. Only `err` is accessible. I need response.headers too. So the only way is to use the callback, and wrap it myself.

pvdlg replied:
Ah good point. For info you can solve that by using [pify with the multiargs](https://github.com/sindresorhus/pify#multiargs) option.
But that would be something for another PR

kbrandwijk replied:
I searched for a lot of libraries, including bluebird and es6-promisify, but they both support multiArgs only on success, not on error. Thank you for pointing me at this, I will create an issue for it.

pvdlg commented on src/lib/npm.js:
```diff
@@ -25,22 +22,65 @@ async function getNpmToken({npm, options}) {
+    return false;
+  }
+
+  return new Promise(resolve => {
```
Same here, you could use `async/await` for consistency and clarity.

kbrandwijk replied:
Here, the same thing applies. There are cases where no error is returned, but response.body.ok == false. I can't check that without access to response, which again requires the callback.
